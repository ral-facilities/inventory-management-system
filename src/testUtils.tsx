import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { RenderOptions, render } from '@testing-library/react';
import { enGB } from 'date-fns/locale/en-GB';
import React from 'react';
import {
  RouterProvider,
  createBrowserRouter,
  createMemoryRouter,
} from 'react-router-dom';
import { paths } from './App';
import {
  CatalogueCategory,
  CatalogueCategoryProperty,
  CatalogueItem,
  Item,
  Manufacturer,
} from './api/api.types';
import CatalogueCategoriesJSON from './mocks/CatalogueCategories.json';
import CatalogueItemsJSON from './mocks/CatalogueItems.json';
import ItemsJSON from './mocks/Items.json';
import ManufacturersJSON from './mocks/Manufacturers.json';

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
  });

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  queryClient?: QueryClient;
}

function constructRouterProvider(
  ui: React.ReactNode,
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const Root: React.FunctionComponent = () => {
    return (
      <LocalizationProvider adapterLocale={enGB} dateAdapter={AdapterDateFns}>
        <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
      </LocalizationProvider>
    );
  };

  const router =
    initialEntry !== undefined && urlPathKey !== undefined
      ? createMemoryRouter([{ path: paths[urlPathKey], Component: Root }], {
          initialEntries: [initialEntry],
        })
      : createBrowserRouter([{ path: '*', Component: Root }]);

  return { router: router, provider: <RouterProvider router={router} /> };
}

function constructRouterProviderWrapper(
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const wrapper = ({
    children,
  }: {
    children: React.ReactNode;
  }): JSX.Element => {
    return constructRouterProvider(
      children,
      queryClient,
      urlPathKey,
      initialEntry
    ).provider;
  };
  return wrapper;
}

export function renderComponentWithRouterProvider(
  ui: React.ReactElement,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string,
  {
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const { router, provider } = constructRouterProvider(
    ui,
    queryClient,
    urlPathKey,
    initialEntry
  );
  return {
    queryClient,
    router,
    ...render(ui, {
      wrapper: () => provider,
      ...renderOptions,
    }),
  };
}

export const hooksWrapperWithProviders = (props?: {
  queryClient?: QueryClient;
  urlPathKey?: keyof typeof paths;
  initialEntry?: string;
}) => {
  const testQueryClient = props?.queryClient ?? createTestQueryClient();
  return constructRouterProviderWrapper(
    testQueryClient,
    props?.urlPathKey,
    props?.initialEntry
  );
};

export const getCatalogueItemsPropertiesById = (
  id: string
): CatalogueCategoryProperty[] => {
  const filteredCategories = CatalogueCategoriesJSON.filter(
    (catalogueCategory) => catalogueCategory.id === id
  );

  if (filteredCategories.length === 0) {
    return [];
  }

  const properties = filteredCategories[0].properties ?? [];
  return properties as CatalogueCategoryProperty[];
};

export const getCatalogueCategoryById = (
  id: string
): CatalogueCategory | undefined => {
  return (
    (CatalogueCategoriesJSON.find(
      (catalogueCategory) => catalogueCategory.id === id
    ) as CatalogueCategory) || undefined
  );
};

export const getCatalogueItemById = (id: string): CatalogueItem | undefined => {
  return (
    (CatalogueItemsJSON.find(
      (catalogueItem) => catalogueItem.id === id
    ) as CatalogueItem) || undefined
  );
};

export const catalogueItemData = (id: string): CatalogueItem[] => {
  return CatalogueItemsJSON.filter(
    (catalogueItem) => catalogueItem.catalogue_category_id === id
  );
};

export const getManufacturerById = (id: string): Manufacturer | undefined => {
  return (
    (ManufacturersJSON.find(
      (manufacturer) => manufacturer.id === id
    ) as Manufacturer) || undefined
  );
};

export const getItemsByCatalogueItemId = (catalogueItemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.catalogue_item_id === catalogueItemId);
};

export const getItemById = (itemId: string): Item | undefined => {
  return ItemsJSON.find((item) => item.id === itemId);
};

export const getItemsBySystemId = (systemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.system_id === systemId);
};

export const CREATED_MODIFIED_TIME_VALUES = {
  created_time: '2024-01-01T12:00:00.000+00:00',
  modified_time: '2024-01-02T13:10:10.000+00:00',
};
