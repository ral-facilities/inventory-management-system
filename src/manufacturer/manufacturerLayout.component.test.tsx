import { QueryClient } from '@tanstack/react-query';
import { screen, waitFor } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import type { LoaderFunctionArgs } from 'react-router';
import { renderComponentWithRouterProvider } from '../testUtils';
import ManufacturerLayout, {
  ManufacturerLayoutErrorComponent,
  manufacturerLayoutLoader,
} from './manufacturerLayout.component';

const mockedUseNavigate = vi.fn();

vi.mock('react-router', async () => ({
  ...(await vi.importActual('react-router')),
  useNavigate: () => mockedUseNavigate,
}));

describe('Manufacturer Layout', () => {
  let user: UserEvent;
  const createView = (path: string, isLandingPage?: boolean) => {
    return renderComponentWithRouterProvider(
      <ManufacturerLayout />,
      isLandingPage ? 'manufacturer' : 'manufacturers',
      path
    );
  };
  beforeEach(() => {
    user = userEvent.setup();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('navigates back to the root directory from root', async () => {
    createView('/manufacturers');

    await waitFor(() => {
      expect(screen.queryByText('Manufacturer A')).not.toBeInTheDocument();
    });

    const homeButton = screen.getByRole('button', {
      name: 'navigate to manufacturers home',
    });

    await user.click(homeButton);

    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
    expect(mockedUseNavigate).toHaveBeenCalledWith('/manufacturers');
  });
  it('navigates back to the root directory from landing page', async () => {
    createView('/manufacturers/1', true);

    await waitFor(() => {
      expect(screen.queryByText('Manufacturer A')).not.toBeInTheDocument();
    });

    const homeButton = screen.getByRole('button', {
      name: 'navigate to manufacturers home',
    });

    await user.click(homeButton);

    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
    expect(mockedUseNavigate).toHaveBeenCalledWith('/manufacturers');
  });
});

describe('Manufacturer Layout Error Component', () => {
  let user: UserEvent;

  beforeEach(() => {
    user = userEvent.setup();
  });
  const createView = () => {
    return renderComponentWithRouterProvider(
      <ManufacturerLayoutErrorComponent />
    );
  };

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders manufacturer error page correctly', async () => {
    const view = createView();

    await waitFor(() => {
      expect(
        screen.getByRole('button', {
          name: 'navigate to manufacturers home',
        })
      ).toBeInTheDocument();
    });

    const homeButton = screen.getByRole('button', {
      name: 'navigate to manufacturers home',
    });

    await user.click(homeButton);

    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
    expect(mockedUseNavigate).toHaveBeenCalledWith('/manufacturers');

    expect(view.asFragment()).toMatchSnapshot();
  });
});

describe('manufacturerLayoutLoader', () => {
  let queryClient: QueryClient;

  beforeEach(() => {
    queryClient = new QueryClient();
    vi.clearAllMocks();
  });
  it('should fetch catalogue category data if catalogue_category_id is provided', async () => {
    const params = { manufacturer_id: '1' };
    const output = await manufacturerLayoutLoader(queryClient)({
      params,
    } as unknown as LoaderFunctionArgs);

    expect(output).toEqual(params);
  });

  it('should throw an error if an invalid catalogue_category_id is provided', async () => {
    const params = { manufacturer_id: '120' };

    await expect(
      manufacturerLayoutLoader(queryClient)({
        params,
      } as unknown as LoaderFunctionArgs)
    ).rejects.toThrow('Request failed with status code 404');
  });
});
