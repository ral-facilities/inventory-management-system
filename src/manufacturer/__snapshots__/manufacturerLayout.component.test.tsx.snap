// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Manufacturer Layout Error Component > renders manufacturer error page correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-v2kfba"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1r501zp-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-70qvj9"
      >
        <span
          aria-label="Manufacturers Home"
          class=""
          data-mui-internal-clone-element="true"
        >
          <button
            aria-label="navigate to manufacturers home"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HomeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            >
              <span
                class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
                />
              </span>
            </span>
          </button>
        </span>
        <nav
          aria-label="breadcrumb"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1mc59t7-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          />
        </nav>
      </div>
    </div>
    <div
      class="MuiBox-root css-7gabc3"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-11gsi37-MuiTypography-root"
      >
        Invalid Manufacturer Route
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
      >
        The manufacturer route you are trying to access doesn't exist. Please click the Home button to navigate back to the Manufacturer Home page.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
